### DISCUSSION TITLE
Build a lab with Pymeasure?

### DISCUSSION THREAD

**Discussion by gaelgrissonnanche** [pymeasure/pymeasure#1055] [Q&A]
Hi,
I am building a lab and would like to know if this is a good long-term strategy for using PyMeasure for this development.
The package has been in development for a long time, and the number of active contributors is not that high. I am always worried about packages that are not properly maintained or on this slippery slope. What would you say is the vitality of this package?
However, your work is amazing and PyMeasure would fulfill all my lab's present needs, it is the future I am worried about.

**Comment by BenediktBurger** [comment]
Welcome on the pymeasure repository!

I understand your worries, as we share them. They were exactly the reason I came to pymeasure myself.

Pymeasure has two (basically independent) parts: An instrument driver library and a data acquisition framework. Which part do you intend to use?

Now to your questions:
- pymeasure is labeled "in development", but has been stable for several years. There are several labs using pymeasure and users (obviously using pymeasure) contribute a bunch of new instruments each year.
- We know, that pymeasure is used by many users in labs, therefore we ensure backward compatibility and raise warnings before removing code or doing breaking changes.
- The instrument core code is quite small, so maintenance does not require many contributors. The instrument drivers are basically definitions and do not change that much.
- Last but not least, we have a test framework to ensure that the code keeps working as expected, even during core changes and for new dependencies / python versions.

**Reply by gaelgrissonnanche** [reply]
Thank you very much for your quick answer and for your clear explanation!

I am interested in both! The instrument driver library is very neat, and I could see myself contributing to this nice project, as I have several instruments not on the list. 

But I was especially interested in the data acquisition framework. I am going to play with it and maybe adopt it. :)

**Reply by BenediktBurger** [reply]
Try the data acquisition framework.
For some use cases, it is not adequate.
I guess, that most users use the drivers part.
I myself have a different acquisition setup, which we are in the process of formalizing (LECO, with the python package pyleco).
If you are looking for a more plug'n'play GUI alternative, there is pymodaq.

**Reply by gaelgrissonnanche** [reply]
Ah I see, well that is a very useful info and thank you for these two frameworks!
I really thought people were primarily using the data acquisition framework. And that is especially this part I was worried about regarding maintenance. I played with it yesterday, it seemed nice, there are really good ideas here, I wonder however if that will correspond to my needs, as I need to monitor experiments for a very long time, over several days and might need to interact remotely from time to time with the instruments during the experiment.

As you seem very knowledgeable, and if you don't me asking: I would like to continuously read all the instruments during the experiment and save them in a log file, and on top of that there would be a routine changing periodically parameters of some instruments, that seems to be a very simple design in theory. I come from Labview, and the thing I don't know how to do is manage complex threads and queues in Python. I saw that there was an Actor Model package for Python, Pykka, that takes care of the low level threading, do you recommend this kind of things? Or do you recommend to stick to managing Thread, Events and Queues ourselves in an experiment? I feel that is the most cumbersome part of the development, but I might be wrong.

**Reply by seb5g** [reply]
If I may jump in the discussion as PyMoDAQ's main developer (and pymeasure contributor ;-) ) . PyMoDAQ has a dedicated extension to do Data logging (with plotting, saving being builtin). At any time during your experiment, you can change whatever parameter you want in the Control Module DashBoard (linked to whatever experiment parameter instrument you wish). And this change can even be done remotely using the legacy TCP/IP communication or in the near future the pyleco featured control modules (we worked together with @BenediktBurger  to bring this into pymodaq). So again if I may, you should have a look there: http://pymodaq.cnrs.fr/en/latest/

**Reply by BenediktBurger** [reply]
Pymodaq is a GUI framework which does all the threads stuff for you.
leco is a communication protocol, which allows to separate your setup into different parts, such that you do not have to worry about threads either.
For example, there is one process with all the instrument drivers in their own thread. Another program (with GUI if you like) acquiring the data from the instruments. An last but not least, one or more programs sending commands to the instrument drivers changing parameters.
This last program can also control the data logger (starting a new measurement, saving one, etc.).

LECO, however only provides the communication tools, and requires some python work on your side.
Here's a description of my setup: https://github.com/pymeasure/pyleco/discussions/46 

Maybe pymodaq is better suited for you (or a mix).

**Reply by gaelgrissonnanche** [reply]
Thank you very much for this very precious info! I will check PyMODAQ, it seems very good.
And LECO seems as much interesting, I see that you are using Actors to manage the threads, that makes sense, that is such a convenient paradigm. Did you every try Pykka? Or did you write your own Actor model?

**Reply by BenediktBurger** [reply]
I didn't know pykka. We developed the actors for LECO.

**Comment by msmttchr** [comment]
Hello Gael,

I had the same concern when I did the selection few years ago for my
department in my company.
We came from an experience where everything was python based and almost
everything was done in house.
We had other people using labview, but this was never a good option for our
needs.

After few years of experience with pymeasure, I have to say that selection
was good, but, like all the open source projects require a bit of
involvement.

We use pymeasure components as:
- instrument interface (relying on PyVISA)
- GUI to write experiments with graphical input, sequencer and live
plotting (via PyQt, pyqtgraph)
- Batch to run experience in batch mode

Our main concern was to be able to express in a concise way an experiment
setup and description such that can be repeated and shared with other
people.

Some limitations, in my opinion, are:
- instrument: lack of generic class to allow to describe experiments
without being linked to a specific brand/model of instrument (there is work
in progress in this area)
- live plotting: x/y plotting limited to one y value (this is being
addressed in pyqtgraph)

In summary, I am a satisfied user of complete pymeasure features, I like
the open source approach and I am clear that this is not a commercial tool,
so some initiative/effort is required to push forward some features of your
interest.

Hope this helps.

Michele

Il giorno dom 3 mar 2024 alle 14:36 Gael Grissonnanche <
***@***.***> ha scritto:

**Reply by gaelgrissonnanche** [reply]
Thank you so much for your detailed feedback. I'll make a note of it, that is very good to know!

**Comment by bklebel** [comment]
This discussion is already a bit old, but I would still like to chime in: 

In my opinion, pymeasure (or at least parts of it) is a good choice for a research lab: 

- It requires minimal effort to use existing instrument drivers, and quite little effort to implement new ones - the entry threshold for the driver library is small. 
- It is kept in pure python, so students do not need to learn a host of programming languages to work with it, but a single one, which is very popular in a range of fields
- The development is ongoing, just as research is. pymeasure (as I see it) has been primarily maintained by PhD students who work with it in their labs, as well as some people in the private industry, where the company (apparently) thankfully decided that some of the time of the employees may go to Open Source projects. Being "in development" in this case (I think) is a good thing, as it demonstrates being maintained, and improved according to the needs of the research labs. We just don't come around to say "this is now the completed version Y.X" very often. 
- By using pymeasure, and having students contributing to it, you can have students learn and teach python, and you do not have to rely of a homebrew solution of a single student, which might be way less maintainable than this collaborative effort. By allowing students to contribute (and possibly help in the maintenance), you can effectively contribute to keeping pymeasure alive, as you put it. This will also help students in their future career, as they learn how Open Source software collaboration can work. 
- In most research groups, a small number of personnel does much of the software development workload - by collaborating with other such (almost) individuals in other groups, all of them gain a lot, and the whole research lab may benefit from the knowledge of other groups' members, collaboration at its best
- LECO is very promising as an overall communication standard, and pyleco will allow for simple integration of pymeasure instrument drivers with a bigger data acquisition system, as was described above for pymodaq. By learning basics of instrument control through using pymeasure, students may have a good start into the topic. 
- We have quite motivated maintainers: you will notice, that although there are quite a few Pull Requests open, this is not due to a lack of interaction (and thus non-merging) from the side of the maintainers, but because they diligently review the valuable contributions, and often there are a number of points to address before merging. 
- Compared to PyMoDAQ, pymeasure is a little smaller, both in possibilities and the entry threshold. If you aim for a synchrotron facility, pymeasure is not quite the tool for you. If you aim for a lab of a few rooms, with a number of different individual setups, a tool like PyMoDAQ (and other synchrotron-level tools) will be overkill, and pymeasure/pyleco might be a good point to start. 

In turn, research labs can contribute a lot to science in general, by using and contributing to pymeasure (or similar packages), and consequently promoting a collaborative way of conducting scientific research. There is an incredible amount of competition and (hidden) hostility in science (in some fields more than in others), so some collaboration on points which are unrelated to research topics in themselves, are surely helpful to science. 
Some of the arguments which I here put forth will apply to a number of different python data acquisition packages as well - however, in my experience, without personnel which is fully dedicated to research software engineering, a low entry threshold is very important, and pymeasure does a good job on this point. 

In conclusion, all the best for the new lab :)
